### PR TITLE
nvidia-container-toolkit: only mount existing paths in the host

### LINKED ISSUE #319201: nvidia-container-toolkit-cdi-generator should not mount nvidia-powerd when nvidia driver == 470
### Describe the bug
The `nvidia-container-toolkit-cdi-generator` service will ask docker to mount the binary `nvidia-powerd` which is not provided by the Nvidia driver v470. 

https://github.com/NixOS/nixpkgs/blob/47b604b07d1e8146d5398b42d3306fdebd343986/nixos/modules/services/hardware/nvidia-container-toolkit/default.nix#L95

Using NixOS 24.05 with nixpkgs `47b604b0`.

### Steps To Reproduce

```nix
      nixpkgs.config.nvidia.acceptLicense = true;
      hardware.nvidia = {
        package = config.boot.kernelPackages.nvidiaPackages.legacy_470;
        modesetting.enable = false;
        powerManagement.enable = true;
        powerManagement.finegrained = false;
        open = false;
        nvidiaSettings = true;
      };
      hardware.opengl = {
        enable = true;
        driSupport = true;
        driSupport32Bit = true;
      };
      hardware.nvidia-container-toolkit.enable = true;
      services.xserver.videoDrivers = [ "nvidia" ];
```

### Expected behavior
`nvidia-powerd` should not be mounted when the binary is missing.

### Screenshots
image

image

### Notify maintainers
@ereslibre

### Metadata

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.32, NixOS, 24.05 (Uakari), 24.05.20240611.47b604b`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.2`
 - nixpkgs: `/nix/store/0vpk9v6rj551q42ym4h9qrdx9xxxhcsg-source`
```
---

Edited: a workaround

```nix
      systemd.services.nvidia-container-toolkit-cdi-generator = {
        path = [ pkgs.jq pkgs.moreutils ];
        postStart = ''
         jq '."containerEdits"."mounts" |= map(select(."containerPath" != "/usr/bin/nvidia-powerd"))' /run/cdi/nvidia-container-toolkit.json | sponge /run/cdi/nvidia-container-toolkit.json
        '';
      };
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by ereslibre**
## Description of changes

Although the final fix for this kind of issues is https://github.com/NixOS/nixpkgs/issues/290609 (`unbreak: we're currently mounting incomplete closures; need to use exportReferencesGraph.`), this adds yet another fix for a symptom.

Fixes: https://github.com/NixOS/nixpkgs/issues/319201

I will address the `exportReferencesGraph` soon that I expect will fix all these problems once and for all, but fix this problem in the meantime.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by SomeoneSerge** [comment]
@ereslibre may I ask you to instead consider whether it's possible to write a NixOS test covering all these symptoms we've had to address so far? E.g. maybe we could manually prepare a fake CDI json, and create temporary fake files in place of device nodes and userspace drivers, and then run a podman/docker container with a script ensuring that all of these fake paths are mounted correctly, including referential integrity (`readlink -f /run/opengl-driver/lib/libcuda.so`, etc)

This would be very useful for our migration to `exportReferencesGraph`...

**Comment by ereslibre** [comment]
> @ereslibre may I ask you to instead consider whether it's possible to write a NixOS test covering all these symptoms we've had to address so far?

I think this is a good idea. I have never written NixOS tests but I will take a look at existing ones and update the PR.

**Comment by SomeoneSerge** [comment]
@ereslibre actually, I now think I shouldn't have delayed merging this; we'll still need the tests though

File: nixos/modules/services/hardware/nvidia-container-toolkit/cdi-generate.nix [hunk 1]
```diff
@@ -13,10 +13,10 @@
     options = mountOptions;
   };
   jqAddMountExpression = ".containerEdits.mounts[.containerEdits.mounts | length] |= . +";
-  allJqMounts = lib.concatMap
+  existingJqMounts = lib.concatMap
     (mount:
       ["${lib.getExe jq} '${jqAddMountExpression} ${builtins.toJSON (mkMount mount)}'"])
-    mounts;
+    (builtins.filter (mount: builtins.pathExists mount.hostPath) mounts);
```
Comment (SomeoneSerge) [inline nixos/modules/services/hardware/nvidia-container-toolkit/cdi-generate.nix:19]: Are you sure you want `builtins.pathExists` at eval time and the eval machine, rather than at runtime at the host machine? 🙃

**Comment by ereslibre** [comment]
> @ereslibre actually, I now think I shouldn't have delayed merging this; we'll still need the tests though

Tests are WIP (https://github.com/ereslibre/nixpkgs/compare/nvidia-ctk-mount-existing-paths-only...ereslibre:nixpkgs:nvidia-ctk-mount-existing-paths-only-wip); I'd like to finish them before 29th june. Otherwise I fear I'll have to resume second half of July. The "referential integrity" at this time is pretty "poor man"; not sure if you had any other approach in mind.

File: nixos/modules/services/hardware/nvidia-container-toolkit/cdi-generate.nix (hunk 1 above)
Comment (ereslibre) [reply nixos/modules/services/hardware/nvidia-container-toolkit/cdi-generate.nix:19]: That's actually a good point. Since we cannot presume that the user will provide mounts through nix paths, they could do:

hardware.nvidia-container-toolkit.mounts = [ ... { hostPath = "/some/path"; containerPath = "/some/path"; } ... ]

We cannot ensure that this will available in runtime at the nix store, so we want to perform this check on runtime, you are right. Let me check.

Comment (SomeoneSerge) [reply nixos/modules/services/hardware/nvidia-container-toolkit/cdi-generate.nix:19]: Yes but more importantly the machine running `nixos-rebuild` (and doing the evaluation) is not necessarily the one on which you deploy the built system (`nixos-rebuild ... --target-host ...`), and things may change between eval time and deployment time.

TLDR: You do need to do the filtering at runtime, in a script similar to have you run jq

File: nixos/modules/services/hardware/nvidia-container-toolkit/cdi-generate.nix
Comment (SomeoneSerge) [inline nixos/modules/services/hardware/nvidia-container-toolkit/cdi-generate.nix:1]: Oh right, we haven't followed up on this one. Do we want this merged or is it intional this is kept as a draft? IIRC we postponed merging because we wanted to add the exportReferencesGraph integration?

Comment (ereslibre) [reply nixos/modules/services/hardware/nvidia-container-toolkit/cdi-generate.nix:1]: I am streamlining all the changes and cleaning up. I expect to remove the draft status tomorrow, thanks a lot for following up.

Comment (ereslibre) [reply nixos/modules/services/hardware/nvidia-container-toolkit/cdi-generate.nix:1]: That said: if you have comments already I'm happy to make changes now based on feedback :)

File: nixos/modules/services/hardware/nvidia-container-toolkit/cdi-generate.nix [hunk 2]
```diff
@@ -32,6 +35,18 @@ function cdiGenerate {
     --nvidia-ctk-path ${lib.getExe' nvidia-container-toolkit "nvidia-ctk"}
 }
 
-cdiGenerate | \
-  ${lib.concatStringsSep " | " allJqMounts} > $RUNTIME_DIRECTORY/nvidia-container-toolkit.json
+function additionalMount {
```
Comment (SomeoneSerge) [inline nixos/modules/services/hardware/nvidia-container-toolkit/cdi-generate.nix:38]: Well one comment is that this could be rewritten with `pythonMinimal` or a similar language that doesn't try so much to make this painful

Comment (SomeoneSerge) [reply nixos/modules/services/hardware/nvidia-container-toolkit/cdi-generate.nix:38]: But as commented before, if you say this is the current iteration I'll go with it

File: nixos/tests/nvidia-container-toolkit.nix [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+            echo "checking $file referential integrity"
+            files=$(set -o pipefail && \
+                    ${pkgs.glibc.bin}/bin/ldd "$1" | \
+                    ${pkgs.gnugrep}/bin/grep '=>' | \
+                    ${pkgs.gnused}/bin/sed "s/.* => //" | \
+                    ${pkgs.gnused}/bin/sed "s/ (.*//") || exit 1
+
+            for file in $files; do
+              if [ ! -f "$file" ]; then
+                die "$file does not exist in the container filesystem"
```
Comment (SomeoneSerge) [inline nixos/tests/nvidia-container-toolkit.nix:34]: I think if a path was reported on the RHS of `=>` it means it was loaded, which means it does exist

File: nixos/tests/nvidia-container-toolkit.nix [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+          check_directory_referential_integrity "/usr/local/nvidia" || exit 1
+        '';
+      in
+      pkgs.dockerTools.buildImage {
+        name = "cdi-test";
+        tag = "latest";
+        config = {
+          Cmd = [ "${testCDIScript}/bin/test-cdi" ];
+          Env = [
+            "LD_LIBRARY_PATH=${unfreeAndInsecurePkgs.linuxPackages.nvidia_x11}/lib:${lib.getLib unfreeAndInsecurePkgs.openssl_1_1}/lib"
```
Comment (SomeoneSerge) [inline nixos/tests/nvidia-container-toolkit.nix:60]: Getting the unfree stuff into the CI is tricky, but thankfully there's no need: you can build the open driver instead

File: nixos/tests/nvidia-container-toolkit.nix [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+            for file in $files; do
+              if [ ! -f "$file" ]; then
+                die "$file does not exist in the container filesystem"
+              fi
+            done
+          }
+
+          check_directory_referential_integrity() {
+            ${pkgs.findutils}/bin/find "$1" -type f -print0 | while read -d $'\0' file; do
+              if [[ $(${pkgs.file}/bin/file "$file" | ${pkgs.gnugrep}/bin/grep ELF) ]]; then
```
Comment (SomeoneSerge) [inline nixos/tests/nvidia-container-toolkit.nix:41]: `getExe`

File: nixos/tests/nvidia-container-toolkit.nix [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+            for file in $files; do
+              if [ ! -f "$file" ]; then
+                die "$file does not exist in the container filesystem"
+              fi
+            done
+          }
+
+          check_directory_referential_integrity() {
+            ${pkgs.findutils}/bin/find "$1" -type f -print0 | while read -d $'\0' file; do
```
Comment (SomeoneSerge) [inline nixos/tests/nvidia-container-toolkit.nix:40]: I usually do the `while ... done < <( subprocess without a subshell )` pattern, although I've been told some people prefer the subshell for whatever reason

File: nixos/tests/nvidia-container-toolkit.nix [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+        };
+    };
+    testScript = ''
+      start_all()
+
+      with subtest("Generate the CDI spec (empty) for a machine with no Nvidia GPU"):
+        no_nvidia_gpus.wait_for_unit("nvidia-container-toolkit-cdi-generator.service")
+        no_nvidia_gpus.succeed("cat /var/run/cdi/nvidia-container-toolkit.json | jq")
+
+      with subtest("Generate the CDI spec for a machine with an Nvidia GPU"):
```
Comment (SomeoneSerge) [inline nixos/tests/nvidia-container-toolkit.nix:151]: Looking at the test, I suppose what it ensures is that `podman` as ran with `--device=...` does read the spec from the expected location... I didn't quite get that idea from the test name

File: nixos/tests/nvidia-container-toolkit.nix [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+            jq
+            podman
+          ];
+          hardware = {
+            nvidia-container-toolkit = {
+              enable = true;
+              package = pkgs.stdenv.mkDerivation {
+                name = "nvidia-ctk-dummy";
+                version = "1.0.0";
+                phases = "installPhase";
```
Comment (SomeoneSerge) [inline nixos/tests/nvidia-container-toolkit.nix:127]: Alternative: `dontUnpack`, `dontBuild`

Comment (SomeoneSerge) [reply nixos/tests/nvidia-container-toolkit.nix:127]: There was a sentiment somewhere that overriding phases like that is not cosher

File: nixos/modules/services/hardware/nvidia-container-toolkit/cdi-generate.nix (hunk 2 above)
Comment (ereslibre) [reply nixos/modules/services/hardware/nvidia-container-toolkit/cdi-generate.nix:38]: Yes, I'll work on a reimplementation of this logic with Python in future PR's, if that's fine with you.

File: nixos/tests/nvidia-container-toolkit.nix (hunk 3 above)
Comment (ereslibre) [reply nixos/tests/nvidia-container-toolkit.nix:34]: Right, I guess I can grep for "not found" then. I don't love this poor man referential integrity check TBH.

File: nixos/tests/nvidia-container-toolkit.nix (hunk 6 above)
Comment (ereslibre) [reply nixos/tests/nvidia-container-toolkit.nix:40]: No special reason, let me know if you think I should change it, but I am fine with both.

File: nixos/tests/nvidia-container-toolkit.nix (hunk 7 above)
Comment (ereslibre) [reply nixos/tests/nvidia-container-toolkit.nix:151]: Yes exactly, and that the JSON is valid etc. I'll name the test in a more descriptive way.

File: nixos/tests/nvidia-container-toolkit.nix (hunk 3 above)
Comment (SomeoneSerge) [reply nixos/tests/nvidia-container-toolkit.nix]: Aah I see, we don't automatically know if a line is a path or a not-found 👍🏻

**Comment by ereslibre** [comment]
@SomeoneSerge: is this good to merge on its current status?

File: nixos/tests/nvidia-container-toolkit.nix [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+    nodes = {
+      no-nvidia-gpus =
+        { config, ... }:
+        {
+          environment.systemPackages = with pkgs; [ jq ];
+          hardware = {
+            inherit nvidia-container-toolkit;
+            nvidia = {
+              open = true;
+              package = config.boot.kernelPackages.nvidiaPackages.stable.open;
```
Comment (SomeoneSerge) [inline nixos/tests/nvidia-container-toolkit.nix:111]: Ehhh it's stupid that `open` doesn't change the default to... an open driver

File: nixos/tests/nvidia-container-toolkit.nix [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+            jq
+            podman
+          ];
+          hardware = {
+            inherit nvidia-container-toolkit;
+            nvidia = {
+              open = true;
+              package = config.boot.kernelPackages.nvidiaPackages.stable.open;
+            };
+            opengl.enable = true;
```
Comment (SomeoneSerge) [inline nixos/tests/nvidia-container-toolkit.nix:130]: I don't understand why Ofborg isn't complaining, we have a `mkRenamedModule` for this one

**Comment by SomeoneSerge** [comment]
I pushed a few cosmetic changes, tell me if they're ok by you

**Comment by ereslibre** [comment]
> I pushed a few cosmetic changes, tell me if they're ok by you

Thank you @SomeoneSerge! LGTM :)

**Comment by ereslibre** [comment]
@SomeoneSerge gentle reminder for merging if everything is fine
